File: lab1/python/main.py
import threading


# Буфер обмена данными для потоков
BUF = ""

def supplier_target (string_with_text = ''):
    """
    функция для потока поставщика
    """
    global BUF

    print('Supplier-thread: start')

    if not string_with_text:
        string_with_text = "Some TEST StrIng.1234#$%^"

    
    BUF = string_with_text
    print('Supplier-thread: fill buffer')

    print('Supplier-thread: end')

def handler_target():
    """
    функция для потока обработчика.
    Принимает ссылку на поток-поставщик, чтобы ожидать его завершения
    """
    global BUF
    
    print("Handler-thread: start")

    BUF = BUF.lower()
    print("Handler-thread: processed the data")

    print("Handler-thread: end")

def consumer_target(pathname):
    """
    функция для потока потребителя.
    Принимает ссылку на поток-обработчик, чтобы ожидать его завершения
    Принимает название файла, куда будет записана информация
    """
    global BUF

    print("Consumer-thread: start")

    with open(pathname,'w')as f:
        f.write(BUF)
    print("Consumer-thread: wrote data")

    print("Consumer-thread: end")


if __name__ == "__main__":
    res_pathname = './results'

    # объявляем потоки
    supplier_thread = threading.Thread(target=supplier_target, args=())
    handler_thread = threading.Thread(target=handler_target, args=())
    consumer_thread = threading.Thread(target=consumer_target, args=(res_pathname,))

    
    supplier_thread.start() # запуск потока
    supplier_thread.join() # ждем завершения потока

    handler_thread.start()
    handler_thread.join()

    consumer_thread.start()
    consumer_thread.join()

    
    
    
    

File: lab1/C/main.c
#include <pthread.h>
#include <stdio.h>
#include<stdlib.h>
#include <string.h>

// буфер обмена данными между потоками
char *BUF;

// функция, которую выполняет поток поставщик данных
// должен получать строку(ASCII), которую запишет в BUF
void *supplier_start(void *text_string){
    printf("Supplier-thread: start\n");

    if (text_string == NULL){
        // Запись тестовой строки в BUF
        char *test_str = "Some TEST StrIng.1234#$^P*";

        // printf("%ld\n", strlen(test_str));
        BUF = malloc(strlen(test_str)*sizeof(char));
        strcpy(BUF, test_str);
        

    }
    else{
        BUF = malloc(sizeof(text_string));
        strcpy(BUF, (char *)text_string);
    }
    printf("Supplier-thread: fill buffer\n");

    printf("Supplier-thread: end\n");    
}

// функция, которую выполняет поток обработчик данных
// Не должен получать ничего
void *handler_start(void *params){
    printf("Handler-thread: start\n");

    int n = strlen(BUF);

    // работает только для ASCII
    for (int i =0; i< n; ++i){
        if ((int)BUF[i] < 91 && (int)BUF[i] > 64){
            BUF[i] = (char)((int)BUF[i] + 32);
        }
    }

    printf("Handler-thread: processed the data\n");

    printf("Handler-thread: end\n");
}

// функция, которую выполняет поток потребитель данных
// Получает строку с путем к файлу, куда запишет BUF
void *consumer_start(void *pathname){
    printf("Consumer-thread: start\n");

    FILE *res;
    res = fopen((char *)pathname,"w+" );
    fprintf(res, "%s", BUF);
    fclose(res);

    printf("Consumer-thread: wrote data\n");

    printf("Consumer-thread: end\n");
}

int main(){

    char *path = "./results";

    pthread_t supplier_tid; // идентификатор потока
    pthread_attr_t supplier_attr; // атрибуты потока    
    pthread_attr_init(&supplier_attr); // получаем дефолтные значения атрибутов


    pthread_t handler_tid;
    pthread_attr_t handler_attr;
    pthread_attr_init(&handler_attr);


    pthread_t consumer_tid;
    pthread_attr_t consumer_attr;    
    pthread_attr_init(&consumer_attr);

    pthread_create(&supplier_tid,&supplier_attr,supplier_start,NULL); // создаем новый поток
    pthread_join(supplier_tid,NULL); //ждем завершения исполнения потока

    pthread_create(&handler_tid,&handler_attr,handler_start,NULL); 
    pthread_join(handler_tid,NULL);

    pthread_create(&consumer_tid,&consumer_attr,consumer_start,path);
    pthread_join(supplier_tid,NULL);
    
    
    free(BUF);
    return 0;
}